File: android/app/src/main/java/sol/vortexx/android/crypto/api/Aead.kt
package sol.vortexx.android.crypto.api

/**
 * Authenticated Encryption with Associated Data.
 *
 * Single responsibility: transform plaintext ↔ packed ciphertext under a
 * symmetric key. The wire format (`nonce || ciphertext || tag`) is fixed
 * across implementations so swapping AES-GCM for ChaCha20-Poly1305 later
 * only touches the impl, never the caller — matches the Vortex server's
 * `ciphertext_hex = nonce(12) + ct + tag(16)` format.
 */
interface Aead {
    /** Encrypt `plaintext` returning `nonce || ciphertext || tag`. */
    fun encrypt(key: ByteArray, plaintext: ByteArray, aad: ByteArray = EMPTY_AAD): ByteArray

    /**
     * Decrypt a `nonce || ciphertext || tag` blob produced by [encrypt].
     *
     * @throws AeadAuthenticationException if the tag does not verify.
     */
    fun decrypt(key: ByteArray, packed: ByteArray, aad: ByteArray = EMPTY_AAD): ByteArray

    companion object { val EMPTY_AAD: ByteArray = ByteArray(0) }
}

class AeadAuthenticationException(message: String, cause: Throwable? = null) : RuntimeException(message, cause)

File: ios/Modules/Sources/VortexCrypto/api/Aead.swift
import Foundation

/// Authenticated Encryption with Associated Data.
///
/// Wire format matches the Vortex server's `ciphertext_hex`:
///     nonce(12) || ciphertext || tag(16)
/// so a hex-encoded payload round-trips unchanged between the Python
/// node, the web client, Android (Kotlin `AesGcmAead`), and this impl.
public protocol Aead: Sendable {
    /// Encrypt `plaintext` returning `nonce || ciphertext || tag`.
    func encrypt(key: Data, plaintext: Data, aad: Data) throws -> Data

    /// Decrypt a packed blob produced by [encrypt].
    /// Throws [AeadError.authenticationFailed] on tag mismatch.
    func decrypt(key: Data, packed: Data, aad: Data) throws -> Data
}

public extension Aead {
    func encrypt(key: Data, plaintext: Data) throws -> Data {
        try encrypt(key: key, plaintext: plaintext, aad: Data())
    }
    func decrypt(key: Data, packed: Data) throws -> Data {
        try decrypt(key: key, packed: packed, aad: Data())
    }
}

public enum AeadError: Error, Equatable {
    case invalidKeyLength(Int)
    case ciphertextTooShort
    case authenticationFailed(String)
}
